### ISSUE TITLE
Memory Leak on UI.Playground

### ISSUE THREAD

**Issue by PoisnFang** [graphql-dotnet/server#402]
There is a memory leak when you are on the page for UI.Playground when I run it in my web app. You can download it and test it: [Here](https://github.com/PoisnFang/Poisn.GraphQL)

**Comment by Shane32** [comment]
Memory leak in the browser?  Can you explain?

**Comment by PoisnFang** [comment]
Just run the project and go to the page for playground /ui/playground and watch the memory just continually climb in the Visual studio memory profiler. It just keeps climbing while you are on that page

**Comment by BenjaBobs** [comment]
Playground will by default poll for the schema I think every 2 seconds. This could be the reason, but then the question is what causes the memory to rise on the server.

Asp.net has multiple garbage collection modes, workstation mode and server mode if I remember correctly. If it's configured for workstation mode (I think this is default when compiling with debug) it won't be as aggressive with garbage collection as long as the system has enough memory and is not asking for more. The significance of this is that it might behave differently and be more stable (memory consumption wise) in production.

Of course there might also be a leak, but my point is it'd require investigation to find out if the symptoms are caused by a leak.

**Comment by Shane32** [comment]
I've been battling memory leaks with my graphql server for a few months.  Eventually I came to a couple conclusions:

First, I had a global data loader (e.g. a singleton service), which caching enabled on it.  Data loader caches don't have a limit on memory use, as they are designed to only have a lifetime of the current request (e.g. a scoped service).  Since it was never GC'd, additional queries would even eventually fill up all available ram on the server and crash the process.  Solving this fixed the actual memory leak, along with a few other issues I was having.  Note that if the data loaders were used as described in the documentation, there would have been no problem.

Second, .Net Core always allocates a large percentage of available memory, and then steadies off as memory usage approaches maximum.  Actual memory in use does not equate to memory allocated.  So when looking at Azure logs, "memory working set" indicates how much RAM is in use by the process, whereas "private bytes" or "memory percentage" is a measure of how much memory is allocated.  After the server has warmed up for a few hours, assuming there's no memory leaks, the figures should hold relatively steady.  The "memory working set" will always be climbing, dropping only when the GC runs (which should be often).  It will be obvious from the memory graphs which generation of objects are GC'd.

Here's a snapshot of the "memory working set" for a few hours, peaking at 1.2GB:
![image](https://user-images.githubusercontent.com/6377684/94475869-b4d0ad00-019d-11eb-9114-57645fdc88c1.png)

Here's the "memory percentage used" graph for the same period.  Azure Linux services don't have the "private bytes" data, but the App Service Plan has 3.5GB of RAM and this graph peaks at 90%.
![image](https://user-images.githubusercontent.com/6377684/94476081-fb260c00-019d-11eb-8fc5-12567c9b9de5.png)

I also wrote a program to stress-test my server, making actual graphql requests and logging response times based on the request.  This not only allowed me to monitor memory consumption, but also to implement query caching and optimize database indexes.

So in short, I recommend using a platform to run many queries against the server, and then monitor the server after it has been running for a few hours to see if memory usage has steadied off or not, keeping in mind the difference between "memory working set" and "private bytes".

On a side note, although there are two different garbage collection modes, I did not find that changing the mode made a difference in memory consumption.  It is noteworthy that for servers with a single core (such as my case), the default GC mode is 'workstation', and this same setting is what's suggested to reduce memory usage by various articles online.

**Comment by sungam3r** [comment]
@PoisnFang Your project uses LocalDb. It will take me some time to configure it.

**Comment by sungam3r** [comment]
OK. I installed LocalDB, SSMS, ef global tool, performed migrations and finally ran the example. I set `schema.polling.interval` in Playground settings to 500 to make introspection requests more often:
![изображение](https://user-images.githubusercontent.com/21261007/95661419-e2234080-0b37-11eb-8eb6-09e7597b984f.png)

Now I see this in Task Manager:
![изображение](https://user-images.githubusercontent.com/21261007/95661431-0717b380-0b38-11eb-8130-3d7895379f7a.png)

Let's clarify. GraphQL Server memory consumption is constant and does not grow. At the same time, I notice a constant increase in browser memory. So there is a problem in the browser side?

**Comment by sungam3r** [comment]
So, I'm not a web developer but I was able to get some stats from Chrome memory profiler tool.
1
![изображение](https://user-images.githubusercontent.com/21261007/95661658-b1dca180-0b39-11eb-9e48-88ae19172da5.png)
2
![изображение](https://user-images.githubusercontent.com/21261007/95661668-be60fa00-0b39-11eb-8411-a8d0102114d3.png)
3
![изображение](https://user-images.githubusercontent.com/21261007/95661674-c91b8f00-0b39-11eb-827f-b5bf86aa3553.png)

Yes, indeed, there is an increase in the number of many objects in the heap. I will ask about this in the Playground repo.

**Comment by sungam3r** [comment]
@PoisnFang @BenjaBobs @Shane32 I posted an issue to Playground repo. From my point of view, I don't know how else I can help.

**Comment by PoisnFang** [comment]
@sungam3r Thanks for the detailed run through, I assume we just wait until we hear back.

@Shane32 Your detailed post was alos very helpful, thanks.

My current solution is to just not use playground for now

**Comment by Shane32** [comment]
I would like to note that in some prior versions of GraphQL.NET Server, there was a memory leak with subscriptions not always disposing correctly.  This took a few months to nail down, but has been fixed.  It could be related to this question.

For now, closing this issue as answered and/or an issue with playground.  Please re-open if there are further developments.
